primitives/runtime/src/offchain/pollable.rs

```diff
@@ -0,0 +1,439 @@
+		mem::forget(Rc::clone(&waker));
+		raw_waker(waker)
+	}
+
+	// Wake by value, moving the Arc into the Wake::wake function
+	unsafe fn wake<W: Wake + Send + 'static>(waker: *const ()) {
+		let waker: Rc<W> = Rc::from_raw(waker as *const W);
+		<W as Wake>::wake(waker);
+	}
+
+	// Wake by reference, wrap the waker in ManuallyDrop to avoid dropping it
+	unsafe fn wake_by_ref<W: Wake + Send + 'static>(waker: *const ()) {
+		let waker: ManuallyDrop<Rc<W>> = ManuallyDrop::new(Rc::from_raw(waker as *const W));
+		<W as Wake>::wake_by_ref(&waker);
+	}
+
+	// Decrement the reference count of the Arc on drop
+	unsafe fn drop_waker<W: Wake + Send + 'static>(waker: *const ()) {
+		mem::drop(Rc::from_raw(waker as *const W));
+	}
+
+	RawWaker::new(
+		Rc::into_raw(waker) as *const (),
+		&RawWakerVTable::new(clone_waker::<W>, wake::<W>, wake_by_ref::<W>, drop_waker::<W>),
+	)
+}
+
+/// Block until a given future is resolved.
+///
+/// ## Support
+/// At the time of writing, the only supported futures are ones that:
+/// 1. are immediately ready (i.e. return `Poll::Ready`),
+/// 2. can be driven by the host via the underlying low-level [`PollableId`] API,
+/// 3. are a combination of the above (e.g. `futures::future::join`ed).
+pub fn block_on<T>(future: impl Future<Output = T>) -> T {
+	pin_mut!(future);
+
+	struct WokenUp(RefCell<bool>);
+	impl Wake for WokenUp {
+		fn wake(self: Rc<Self>) {
+			*self.0.borrow_mut() = true;
+		}
+	}
+
+	let woken_up = Rc::new(WokenUp(RefCell::new(false)));
+
+	let waker = woken_up.clone().into_waker();
+	let mut context = Context::from_waker(&waker);
+
+	loop {
+		// We poll the future continuously until it is either Ready,
+		// or the waker stops being invoked during the polling.
+		loop {
+			match future.as_mut().poll(&mut context) {
+				Poll::Ready(value) => return value,
+				// If the waker has been used during the polling of this future,
+				// then we have to poll again.
+				Poll::Pending if mem::replace(&mut *woken_up.0.borrow_mut(), false) => continue,
+				// Otherwise, we need to wait for an external notification for
+				// this future to make progress.
+				Poll::Pending => break,
+			}
+		}
+
+		let mut state = STATE.lock();
+		debug_assert_eq!(state.message_ids.len(), state.wakers.len());
+
+		// `block` indicates whether we should block the thread or just peek.
+		// Always `true` during the first iteration, and `false` in further iterations.
+		let mut block = true;
+
+		// We process in a loop all pending messages.
+		while let Some((message_id, index_in_list)) = next_notification(&mut state.message_ids, block) {
+			block = false;
+
+			state.message_ids.remove(index_in_list);
+			let waker = state.wakers.remove(index_in_list);
+			waker.wake();
+
+			let _new_msg_id = state.pending_messages.insert(message_id);
+			debug_assert!(_new_msg_id);
+		}
+
+		debug_assert!(!block);
+	}
+}
+
+lazy_static::lazy_static! {
+	// TODO: we're using a Mutex, which is ok for as long as WASM doesn't have threads
```

Comment:
In this module we have a bunch of unsafe code anyway, and the assumption that there is no threading (i.e. no `Sync` requirement) in WASM is already present in some other parts of the code, so might be worth to have a common way (one struct) which encapsulates this assumption and allows restricted mutable access to a global state like that with a tiny bit of unsafe block?